lib/auth/userloginstate/generator.go

```diff
@@ -162,43 +161,55 @@ func (g *Generator) Generate(ctx context.Context, user types.User) (*userloginst
 
 	if g.usageEvents != nil {
 		// Emit the usage event metadata.
-		if err := g.emitUsageEvent(ctx, user, uls); err != nil {
+		if err := g.emitUsageEvent(ctx, user, uls, inheritedRoles, inheritedTraits); err != nil {
 			g.log.Debug("Error emitting usage event during user login state generation, skipping")
 		}
 	}
 
 	return uls, nil
 }
 
-// addAccessListsToState will added the user's applicable access lists to the user login state.
-func (g *Generator) addAccessListsToState(ctx context.Context, user types.User, state *userloginstate.UserLoginState) error {
+// addAccessListsToState will add the user's applicable access lists to the user login state,
+// returning any inherited roles and traits.
+func (g *Generator) addAccessListsToState(ctx context.Context, user types.User, state *userloginstate.UserLoginState) ([]string, map[string][]string, error) {
 	accessLists, err := g.accessLists.GetAccessLists(ctx)
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, nil, trace.Wrap(err)
 	}
 
-	// Create an identity for testing membership to access lists.
-	identity := tlsca.Identity{
-		Username: user.GetName(),
-		Groups:   user.GetRoles(),
-		Traits:   user.GetTraits(),
-		UserType: user.GetUserType(),
+	accessListHierarchy, err := accesslists.NewHierarchy(ctx, accessLists, g.accessLists, g.accessLists, g.clock)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
 	}
 
+	var allInheritedRoles []string
+	allInheritedTraits := make(map[string][]string)
+
 	for _, accessList := range accessLists {
-		if err := services.IsAccessListOwner(identity, accessList); err == nil {
-			g.grantRolesAndTraits(identity, accessList.Spec.OwnerGrants, state)
+		// Grants are inherited if the user is a member of the access list, explicitly or via inheritance.
+		if membershipKind, err := accessListHierarchy.IsAccessListMember(ctx, user, accessList.GetName()); err == nil && membershipKind != accesslists.MembershipOrOwnershipTypeNone {
```

Comment:
Same question as in the other PR - can this be encapsulated by Hierarchy?

```go
roles, traits, err := accessListHierarchy.GetUserGrants(ctx, user, accessList.GetName())
```

The idea for it is that it encapsulates common functionality you need to do with nested access lists, not just ownership/membership checks.

Reply:
Yeah, could do to move this into the hierarchy interface instead!